From aec94cfc9af6b844ee5c092abb9073910c3b6bf1 Mon Sep 17 00:00:00 2001
From: rugosi <kukacster@gmail.com>
Date: Sun, 17 Jan 2021 00:30:40 +0100
Subject: [PATCH] Creating a timestamped copy of backup.prev

On each call of save_backup(), a timestamped copy if created from
backup.prev: backup.prev_20210117_224015
---
 cinelerra-5.1/cinelerra/mwindow.C | 20 ++++++++++++++++++++
 cinelerra-5.1/cinelerra/mwindow.h |  2 ++
 2 files changed, 22 insertions(+)

diff --git a/cinelerra-5.1/cinelerra/mwindow.C b/cinelerra-5.1/cinelerra/mwindow.C
index 04549fde..68cb1037 100644
--- a/cinelerra-5.1/cinelerra/mwindow.C
+++ b/cinelerra-5.1/cinelerra/mwindow.C
@@ -4270,6 +4270,25 @@ void MWindow::get_backup_path(char *path, int len)
 	cp += snprintf(cp, ep-cp, idx ? BACKUPn_FILE : BACKUP_FILE, idx);
 }
 
+void MWindow::create_timestamped_copy_from_previous_backup(char *previouspath)
+{
+  if (previouspath == nullptr) return;
+  char backup_path[BCTEXTLEN];
+  backup_path[0] = 0;
+  time_t now = time(NULL);
+  struct tm* currenttime = localtime(&now);
+  snprintf(backup_path, sizeof(backup_path), 
+      "%s/%s_%d%.2d%.2d_%.2d%.2d%.2d",
+      File::get_config_path(), BACKUP_FILE1, 
+      currenttime->tm_year + 1900,
+      currenttime->tm_mon + 1,
+      currenttime->tm_mday,
+      currenttime->tm_hour,
+      currenttime->tm_min,
+      currenttime->tm_sec);
+	rename(previouspath, backup_path);
+}
+
 void MWindow::save_backup()
 {
 	FileXML file;
@@ -4280,6 +4299,7 @@ void MWindow::save_backup()
 	snprintf(backup_path1, sizeof(backup_path1), "%s/%s",
 		File::get_config_path(), BACKUP_FILE1);
 	get_backup_path(backup_path, sizeof(backup_path));
+  create_timestamped_copy_from_previous_backup(backup_path1);
 	rename(backup_path, backup_path1);
 	edl->save_xml(&file, backup_path);
 	file.terminate_string();
diff --git a/cinelerra-5.1/cinelerra/mwindow.h b/cinelerra-5.1/cinelerra/mwindow.h
index 8f9bfbb7..0afa1b0c 100644
--- a/cinelerra-5.1/cinelerra/mwindow.h
+++ b/cinelerra-5.1/cinelerra/mwindow.h
@@ -23,6 +23,7 @@
 
 #include <stdio.h>
 #include <stdint.h>
+#include <time.h>
 
 #include "apatchgui.h"
 #include "arraylist.h"
@@ -384,6 +385,7 @@ public:
 	void crop_video(int mode);
 	void update_plugins();
 	void get_backup_path(char *path, int len);
+  void create_timestamped_copy_from_previous_backup(char *previouspath);
 // Call after every edit operation
 	void save_backup();
 	void load_backup();
-- 
2.20.1

